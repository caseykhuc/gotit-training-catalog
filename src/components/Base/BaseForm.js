import React from 'react';
import { Form, Alert } from 'react-bootstrap';
import PropTypes from 'prop-types';

const BaseForm = ({
  onInputChange,
  inputError,
  requestError,
  inputValue,
  fields,
  onKeyDown,
}) => (
    <div onKeyDown={onKeyDown}>
      {fields.map(({ name, type, options }) => (
        <Form.Group controlId={name} key={name}>
          {type !== 'select'
            ? (
              <Form.Control
                type={type}
                name={name}
                value={inputValue[name]}
                placeholder={`Enter ${name}`}
                onChange={onInputChange}
                isInvalid={inputError[name]}
              />
            )
            : (
              <Form.Control
                as="select"
                name={name}
                onChange={onInputChange}
                isInvalid={inputError[name]}
              >
                {options.map(({ name, id }) => <option value={id} key={id}>{`Category ${name}`}</option>)}
              </Form.Control>
            )}
          <Form.Control.Feedback type="invalid">{inputError[name]}</Form.Control.Feedback>
        </Form.Group>
      ))}
      {requestError && <Alert variant="danger">{requestError}</Alert>}
    </div>
  )

BaseForm.propTypes = {
  onInputChange: PropTypes.func.isRequired,
  onKeyDown: PropTypes.func.isRequired,
  inputError: PropTypes.object.isRequired,
  inputValue: PropTypes.object.isRequired,
  fields: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    value: PropTypes.string,
  })),
  requestError: PropTypes.string,
}

export default BaseForm;
