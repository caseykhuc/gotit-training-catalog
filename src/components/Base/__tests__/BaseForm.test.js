
import React from 'react';
import { shallow } from 'enzyme';
import { Alert } from 'react-bootstrap'
import BaseForm from '../BaseForm'

describe('components/BaseForm', () => {
  let props;
  let wrapper;
  let container;
  let alert;
  const update = () => {
    wrapper.update();
    container = wrapper.find('div');
    alert = wrapper.find(Alert);
  };
  const setup = () => {
    wrapper = shallow(<BaseForm {...props} />);
    update();
  };
  beforeEach(() => {
    props = {
      inputValue: {
        username: '',
        password: '',
      },
      inputError: {
      },
      requestError: '',
      fields: [{ name: 'username', type: 'text' }, {
        name: 'password', type: 'password',
      }],
      onAction: jest.fn(),
      onKeyDown: jest.fn(),
      onInputChange: jest.fn(),
    }
  });
  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should invoke onAction when user press any key', () => {
    setup();
    container.props().onKeyDown({ key: 'b' });
    expect(props.onKeyDown).toHaveBeenCalled();
  });
  it('should render Alert when requestError exist', () => {
    props.requestError = 'Error';
    setup();
    expect(alert).toHaveLength(1);
  })
})
