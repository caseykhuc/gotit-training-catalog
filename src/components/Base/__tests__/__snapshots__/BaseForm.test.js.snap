// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/BaseForm should render correctly 1`] = `ShallowWrapper {}`;
